bookserver/routers/discuss.py

```diff
@@ -21,19 +70,162 @@
+    will block
+    """
+    rslogger.debug(f"IN WEBSOCKET {uname=}")
+    username = uname
+    local_users.add(username)
+    await manager.connect(username, websocket)
+    r = aioredis.from_url(os.environ.get("REDIS_URI", "redis://localhost:6379/0"))
+    subscriber = r.pubsub()
+
+    async def my_get_message():
+        return await subscriber.get_message(timeout=1.0)
+
+    await subscriber.subscribe("peermessages")
+
+    # multi_await `Docs and links here <http://www.hydrogen18.com/blog/python-await-multiple.html>`_ acts like
+    # the generic sockets select statement, returning a result from get as soon
+    # as ANY of our async functions have a result available.
+    async with multi_await() as mawait:
+        mawait.add(my_get_message)
+        mawait.add(websocket.receive_text)
+
+        while True:
+            # Wait for something to happen
+            # get returns two lists one of completions and one of failures
+            complete, failures = await mawait.get()
+            # i == 0 --> pubsub message
+            # i == 1 --> websocket message
+            pmess, wsres = complete
+            psfail, wsfail = failures
+
+            if wsfail is not None:
+                rslogger.error(wsfail)
+                # The fail is more than likely a runtime error from a page close or refresh
+                # RuntimeError('Cannot call "receive" once a disconnect message has been received.')
+                manager.disconnect(username)
+                return
+
+            if psfail is not None:
+                rslogger.error(psfail)
+                # probably do not want to return
+
+            # handle message from the pubsub queue
+            if pmess is not None:
+                rslogger.debug(f"{pmess=}")
+                if pmess["type"] == "message":
+                    # This is a message sent into the channel, our stuff is in
+                    # the ``data`` field of the redis message
+                    data = json.loads(pmess["data"])
+                else:
+                    rslogger.error("unknown message type {pmess['type']}")
+                    continue
+                if data["broadcast"]:
+                    await manager.broadcast(data)
+                else:
+                    # because **every** connection runs this same loop
+                    # we only want to send a non-broadcast message if
+                    # it is to ourself.
+                    partner = await r.hget("partnerdb", data["from"])
+                    partner = partner.decode("utf8")
+                    if partner == username:
+                        await manager.send_personal_message(partner, data)
+                        # log the message
+                        await create_useinfo_entry(
+                            UseinfoValidation(
+                                event="sendmessage",
+                                act=f"to:{partner}:{data['message']}",
```

Comment:
A long message will cause a validation exception here -- perhaps send only the first 500 chars of the message?

Reply:
safety first!